Index: brightcove-video-connect.php
===================================================================
--- brightcove-video-connect.php	(revision 260988)
+++ brightcove-video-connect.php	(working copy)
@@ -3,7 +3,7 @@
  * Plugin Name: Brightcove Video Connect
  * Plugin URI: https://wordpress.org/plugins/brightcove-video-connect/
  * Description: A Brightcove™ Connector for WordPress that leverages enhanced APIs and Brightcove™ Capabilities
- * Version: 1.2.4
+ * Version: 1.2.5
  * Author: 10up
  * Author URI: http://10up.com
  * License: GPLv2+
@@ -29,7 +29,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  021.0.2301  USA
  */
 
-define( 'BRIGHTCOVE_VERSION', '1.2.4' );
+define( 'BRIGHTCOVE_VERSION', '1.2.5' );
 define( 'BRIGHTCOVE_URL', plugin_dir_url( __FILE__ ) );
 define( 'BRIGHTCOVE_PATH', dirname( __FILE__ ) . '/' );
 define( 'BRIGHTCOVE_BASENAME', plugin_basename( __FILE__ ) );
@@ -102,14 +102,6 @@
 require_once( BRIGHTCOVE_PATH . 'includes/class-bc-setup.php' );
 require_once( BRIGHTCOVE_PATH . 'includes/class-bc-notification-api.php' );
 
-// Check Brightcove status if is_admin().
-if ( is_admin() ) {
-
-	require_once( BRIGHTCOVE_PATH . 'includes/admin/class-bc-status-warning.php' );
-	new BC_Status_Warning();
-
-}
-
 // Upgrade routine
 $installed = get_option( 'brightcove_version' );
 if ( ! $installed || version_compare( $installed, BRIGHTCOVE_VERSION, '<' ) ) {
Index: readme.txt
===================================================================
--- readme.txt	(revision 260988)
+++ readme.txt	(working copy)
@@ -4,8 +4,8 @@
 Donate link:       https://supporters.eff.org/donate
 Tags:              brightcove, 10up, videos, video
 Requires at least: 4.2
-Tested up to:      4.5.3
-Stable tag:        1.2.4
+Tested up to:      4.6.1
+Stable tag:        1.2.5
 License:           GPLv2 or later
 License URI:       http://www.gnu.org/licenses/gpl-2.0.html
 
@@ -188,6 +188,10 @@
 
 == Changelog ==
 
+= 1.2.5 =
+
+* Enhancement: Removed call to Brightcove status API
+
 = 1.2.4 =
 
 * Fix: Fixed issues with "Insert into Post" button

